packages/Kokkos/src/details/DTK_DetailsTreeTraversal.hpp

```diff
@@ -118,123 +106,51 @@ int query_dispatch( BVH<NO> const *bvh, Predicate const &pred,
-                        AABB const &queryAABB, int queryObjectIdx )
-{
-    // Allocate traversal stack from thread-local memory,
-    // and push NULL to indicate that there are no postponed nodes.
-    Node const *stack[64];
-    Node const **stackPtr = stack;
-    *stackPtr++ = NULL; // push
-
-    // Traverse nodes starting from the root.
-    Node const *node = bvh.getRoot();
-    do
-    {
-        // Check each child node for overlap.
-        Node const *childL = node->children.first;
-        Node const *childR = node->children.second;
-        bool overlapL = ( checkOverlap( queryAABB, childL->bounding_box ) );
-        bool overlapR = ( checkOverlap( queryAABB, childR->bounding_box ) );
-
-        // Query overlaps a leaf node => report collision.
-        if ( overlapL && bvh.isLeaf( childL ) )
-            list.add( queryObjectIdx, bvh.getIndex( childL ) );
-
-        if ( overlapR && bvh.isLeaf( childR ) )
-            list.add( queryObjectIdx, bvh.getIndex( childR ) );
-
-        // Query overlaps an internal node => traverse.
-        bool traverseL = ( overlapL && !bvh.isLeaf( childL ) );
-        bool traverseR = ( overlapR && !bvh.isLeaf( childR ) );
-
-        if ( !traverseL && !traverseR )
-            node = *--stackPtr; // pop
-        else
-        {
-            node = ( traverseL ) ? childL : childR;
-            if ( traverseL && traverseR )
-                *stackPtr++ = childR; // push
-        }
-    } while ( node != NULL );
-}
-
-template <typename NO>
-std::list<std::pair<int, double>>
-within( BVH<NO> const &bvh, Point const &query_point, double radius )
+template <typename NO, typename Predicate>
+std::list<int> spatial_query( BVH<NO> const &bvh, Predicate const &predicate )
 {
-    std::list<std::pair<int, double>> ret;
+    std::list<int> ret;
 
     Stack stack;
 
     Node const *node = bvh.getRoot();
-    double node_distance = 0.0;
-    stack.emplace( node, node_distance );
+    stack.emplace( node );
     while ( !stack.empty() )
     {
-        std::tie( node, node_distance ) = stack.top();
+        node = stack.top();
         stack.pop();
         if ( bvh.isLeaf( node ) )
         {
-            ret.emplace_back( bvh.getIndex( node ), node_distance );
+            ret.emplace_back( bvh.getIndex( node ) );
         }
         else
         {
             for ( Node const *child :
                   {node->children.first, node->children.second} )
             {
-                double child_distance =
-                    distance( query_point, child->bounding_box );
-                if ( child_distance <= radius )
-                    stack.emplace( child, child_distance );
+                if ( predicate( child ) )
+                    stack.emplace( child );
             }
         }
     }
-    ret.sort(
-        []( std::pair<int, double> const &a, std::pair<int, double> const &b ) {
-            return a.second > b.second;
-        } );
+
```

Comment:
so we don't maintain them sorted any more.  that's fine with me, I don't think Boost.Geometry does sort them either.